### PR TITLE
Replace Select2 with SlimSelect

### PR DESCRIPTION
**Pull request by gmq**
## Context

Select2 is difficult to configure and use with newer bundlers based on `esbuild`. Rather than attempting to make it work, this pull request replaces the library with a more up-to-date alternative while attempting to maintain similar usage.

## Breaking Changes

This implementation replaces ActiveAdmin's jQuery events with a MutationObserver. Instead of listening to `has_many_*` events, the MutationObserver watches for new elements and initializes `slim-select` where necessary. It covers most typical use cases, but there may be scenarios that I have not taken into account.

## Changes

I added our standard ESLint configuration to accommodate the use of ES6 syntax in newer files. This may result in a lot of red lines in older files, but since they are not touched often, it shouldn't be a major issue.

## Possible Issues
It may be necessary to fork `slim-select` to fix outstanding issues since the author of the library works sporadically on it. This is alleviated in part by the fact that we don't really need any _new_ features. Once we get it working with no issues, there won't be a need for continued development, and we can contribute the fixes to upstream.

### CODE REVIEW

File: lib/activeadmin_addons.rb [hunk 1]
```diff
@@ -4,7 +4,7 @@ module ActiveadminAddons
   attr_writer :default_select, :datetime_picker_default_options, :datetime_picker_input_format
 
   def default_select
-    return "select2" unless @default_select
+    return "slim-select" unless @default_select
```
Comment (monkeyci) [inline lib/activeadmin_addons.rb:7]: **[rubocop]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Add empty line after guard clause.

File: spec/features/inputs/nested_select_input_spec.rb [hunk 2]
```diff
@@ -165,26 +177,42 @@
...
       end
 
       it "sets general minimum_input_length option", js: true do
-        msg = "Please enter 5 or more characters"
-        on_input_ctx("invoice_region_id") { open_select2_options }
-        expect_select2_result_text_to_eq(1, msg)
-        on_input_ctx("invoice_region_id") { select2_options_container.click }
-        on_input_ctx("invoice_city_id") { open_select2_options }
-        expect_select2_result_text_to_eq(1, msg)
+        msg = "Input is too short"
```
Comment (monkeyci) [inline spec/features/inputs/nested_select_input_spec.rb:185]: ⚠️ **[rubocop]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Useless assignment to variable - msg.

File: spec/features/inputs/slimselect_spec.rb [hunk 3]
```diff
@@ -137,28 +132,28 @@
...
-      expect_select2_options_count_to_eq(0)
+      expect { expect_slimselect_options_count_to_eq(0) }.to raise_error(Capybara::ElementNotFound)
       expect(page).not_to have_selector("select.default-select")
-      expect(page).not_to have_selector("select.select2")
+      expect(page).not_to have_selector("select[data-id]")
     end
   end
 
-  context "when default config is Active Admin's default and select control has select 2 class" do
+  context "when default config is Active Admin's default and select control has slim-select class" do
```
Comment (monkeyci) [inline spec/features/inputs/slimselect_spec.rb:141]: **[rubocop]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Line is too long. [101/100]

File: spec/features/inputs/tags_input_spec.rb [hunk 4]
```diff
@@ -43,25 +45,26 @@
...
+        sleep 0.5
         expect { find(item_id, visible: false) }.to raise_error(Capybara::ElementNotFound)
       end
 
       it "does not allow new items", js: true do
-        expect_select2_choices_count_to_eq(1)
-        fill_select2_input_and_press_return("Not preloaded item")
-        expect_select2_choices_count_to_eq(1)
+        expect_slimselect_options_count_to_eq(3)
+        expect { add_slimselect_option("Not preloaded item") }.to raise_error(Capybara::ElementNotFound)
```
Comment (monkeyci) [inline spec/features/inputs/tags_input_spec.rb:66]: **[rubocop]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Line is too long. [104/100]

File: spec/support/capybara_helpers.rb [hunk 5]
```diff
@@ -21,92 +21,99 @@ def expect_css(css)
...
-    find(".select2-container")
+  def on_nested_ctx(resource_number, &block)
+    within("li.has_many_container fieldset:nth-child(#{resource_number + 1}) ") do
+      block.call
+    end
   end
 
-  def select2_input
-    find(".select2-search__field")
+  def expect_slimselect_options_count_to_eq(count, id=nil)
```
Comment (monkeyci) [inline spec/support/capybara_helpers.rb:34]: **[rubocop]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Surrounding space missing in default value assignment.

File: spec/support/capybara_helpers.rb [hunk 6]
```diff
@@ -21,92 +21,99 @@ def expect_css(css)
...
+    id = slimselect_original_select_id unless id
```
Comment (monkeyci) [inline spec/support/capybara_helpers.rb:35]: **[rubocop]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Use the double pipe equals operator ||= instead.

File: spec/support/capybara_helpers.rb [hunk 7]
```diff
@@ -21,92 +21,99 @@ def expect_css(css)
...
-  def expect_select2_choices_count_to_eq(count)
-    expect(page).to have_css("li.select2-selection__choice", count: count)
+  def click_slimselect_option(option_text, id = nil)
+    slimselect_popover(id).find(".ss-option", text: option_text).click
   end
 
-  def expect_select2_options_count_to_eq(count)
-    expect(page).to have_css("select.select2-hidden-accessible option", count: count)
+  def pick_slimselect_entered_option(item_text, display_name = nil)
+    display_name = item_text unless display_name
```
Comment (monkeyci) [inline spec/support/capybara_helpers.rb:79]: **[rubocop]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Use the double pipe equals operator ||= instead.

File: spec/support/capybara_helpers.rb [hunk 8]
```diff
@@ -21,92 +21,99 @@ def expect_css(css)
...
       "li.nested_level:nth-child(#{result_number})", text: /#{text}/
     )
   end
+
+  def close_slimselect_options(id = nil)
+    slimselect_search_input(id || slimselect_original_select_id).native.send_keys(:escape)
+  end
+
+  def expect_slimselect_popover_is_closed(id = nil)
+    expect(page.document).not_to have_css(".ss-content[data-id=#{id || slimselect_original_select_id}]")
```
Comment (monkeyci) [inline spec/support/capybara_helpers.rb:117]: **[rubocop]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Line is too long. [104/100]

**Review by difernandez — CHANGES_REQUESTED** [review]
Nice! Like how the js of different slim select input types is organized 💯. I leave some comments, let me know what you think. I only reviewed the code, when I have some time I'll do some testing in local and I'll come back if I find something

**Review by ldlsegovia — COMMENTED** [review]
I have manually tested the functionality in the Dummy app and I found these details:

1. Indentation:
    before:
         image

    after:
        image

2. Cleaning button
    before:
        image
     after:
          image


3. Red border:
    before:
        image
    after:
        image


4. Clean value after selecting
    before:
        image
    after:
        image

File: app/inputs/nested_level_input.rb [hunk 9]
```diff
@@ -15,6 +15,7 @@ def render_custom_input
 
   def load_control_attributes
     load_class(@options[:class])
+    load_data_attr(:association, value: association_name)
```
[the remark this answers is not in the document]
Comment (gmq) [reply app/inputs/nested_level_input.rb:18]: I use it for finding the correct nested select

https://github.com/platanus/activeadmin_addons/blob/4ab9e2c0b6321271dd7b3cb1705889ba65705188/app/javascript/activeadmin_addons/inputs/slim-select-nested.js#L51

File: app/javascript/activeadmin_addons/inputs/slim-select-tags.js [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+          if (document.querySelectorAll(`#${itemId}`).length > 0) {
+            return;
+          }
+
+          const hiddenInput = document.createElement('input');
+          hiddenInput.id = itemId;
+          hiddenInput.name = itemName;
+          hiddenInput.value = data.value;
+          hiddenInput.type = 'hidden';
+          hiddenInput.textContent = data.text;
```
[the remark this answers is not in the document]
Comment (gmq) [reply app/javascript/activeadmin_addons/inputs/slim-select-tags.js]: It appends the text next to the input, but I can't remember why I added it so I removed it.

**Review by difernandez — CHANGES_REQUESTED** [review]
One small detail, otherwise it's looking pretty good to me 👌I'll try the branch later today to see if there's anything else. One other thing I just remembered: we should update the docs, there are various mentions of select2 in there, as well as filenames like `select2_search.md`

**Review by difernandez — CHANGES_REQUESTED** [review]
Found a couple things while testing this:
- Display text is being interpreted as html. The part betewen `<>` dissapears and is an html tag in devtools

https://user-images.githubusercontent.com/12057523/225655154-c5a8f535-4f28-4748-9c76-57e182206704.mov


image

- Interactive tag is not working for me 🤔 select appears and disappears immediately

https://user-images.githubusercontent.com/12057523/225657794-87e95e37-b836-42c8-ba05-6a9d2d59caed.mov

**Review by difernandez — APPROVED** [review]
excellent! 👏 🚀

**Review by ldlsegovia — APPROVED** [review]
👏🏻

**Pull request merged**